### PR TITLE
[Launcher]: Support Container Request Timeout Mechanism

### PR DESCRIPTION
**Pull request by yqwang-ms**
1. Build mapping among Task, ContainerRequest and Container
2. Add FrameworkEvent and TaskEvent
3. Support Container Request Timeout Mechanism
4. Remove AntiaffinityAllocationManager

### CODE REVIEW

**Comment by yqwang-ms** [comment]
@sterowang is added to the review. #Closed

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java [hunk 1]
```diff
@@ -753,16 +672,36 @@ private void recoverTransitionTaskStateQueue() {
...
+    LOGGER.logInfo("%s: addContainerRequest with timeout %ss. ContainerRequest: [%s]",
+        taskLocator, containerRequestTimeoutSec, HadoopExtensions.toString(request));
     rmClient.addContainerRequest(request);
-    statusManager.transitionTaskState(taskLocator, TaskState.CONTAINER_REQUESTED);
+    statusManager.transitionTaskState(taskLocator, TaskState.CONTAINER_REQUESTED,
+        new TaskEvent().setContainerRequest(request));
+
+    transitionTaskStateQueue.queueSystemTaskDelayed(() -> {
+      if (statusManager.containsTask(request.getPriority())) {
+        LOGGER.logInfo(
```
Comment (qinchen123) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java]: >logInfo [](start = 15, length = 7)

logwarm?

Comment (yqwang-ms) [reply frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java]: OK

---
In reply to: [158203701](https://github.com/Microsoft/pai/pull/121#discussion_r158203701) [](ancestors = 158203701)

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java [hunk 2]
```diff
@@ -753,16 +672,36 @@ private void recoverTransitionTaskStateQueue() {
...
+            "%s: ContainerRequest cannot be satisfied within timeout %ss, Cancel it and Request again. ContainerRequest: [%s]",
+            taskLocator, containerRequestTimeoutSec, HadoopExtensions.toString(request));
+        removeContainerRequest(taskStatus);
+        statusManager.transitionTaskState(taskLocator, TaskState.TASK_WAITING);
+        addContainerRequest(taskStatus);
```
Comment (qinchen123) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java]: > addContainerRequest(taskStatus); [](start = 7, length = 33)

When time out happend, change the "priority" to the latest  sequential number can resolve this issue ?

Comment (yqwang-ms) [reply frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java]: Already changed, pls check 

---
In reply to: [158204184](https://github.com/Microsoft/pai/pull/121#discussion_r158204184) [](ancestors = 158204184)

File: frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java [hunk 3]
```diff
@@ -753,16 +672,36 @@ private void recoverTransitionTaskStateQueue() {
   private void addContainerRequest(TaskStatus taskStatus) throws Exception {
     String taskRoleName = taskStatus.getTaskRoleName();
     TaskStatusLocator taskLocator = new TaskStatusLocator(taskRoleName, taskStatus.getTaskIndex());
-    ContainerRequest request = setupContainerRequest(taskStatus);
+    Integer containerRequestTimeoutSec = CommonUtils.getRandomNumber(
+        conf.getLauncherConfig().getAmContainerRequestMinTimeoutSec(),
+        conf.getLauncherConfig().getAmContainerRequestMaxTimeoutSec());
```
Comment (qinchen123) [inline frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java]: Only the request to the same node will be timeout.  Than, it is not a big big issue.

Comment (yqwang-ms) [reply frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java]: For node specified request, yes, but for nodelabel specified request, no.

---
In reply to: [158221031](https://github.com/Microsoft/pai/pull/121#discussion_r158221031) [](ancestors = 158221031)

**Pull request merged**